### PR TITLE
sinon: 5.x release and cleanup

### PR DESCRIPTION
**Pull request by 43081j**
This is basically to cleanup the sinon types and sort out the recent 5.x release. A few things have been removed, a few have been tightened or simplified.

**This contains #25656 which means sinon 5.x types will _not_ work with TypeScript below version 2.8**

## Changes

- [x] Fake timers can produce/consume numeric IDs or timer objects (in node)
- [x] Static members have all been merged into one declaration for readability
- [x] `spy.reset()` has been removed
- [x] `stub.reset()` has been added
- [x] `stub.callsFake` no longer restricts fakes to a `void` return type
- [x] `sinon.stub(obj, method, fn)` has been removed
- [x] `setImmediate` and `clearImmediate` added to fake timers
- [x] `requestAnimationFrame` and `cancelAnimationFrame` added to fake timers
- [x] `fakeTimer.nextTick` added
- [x] `fakeTimer.tick` can take a string (e.g. `tick("00:10")`)
- [x] `fakeTimer.runMicrotasks()` added
- [x] `fakeTimer.runToFrame()` added
- [x] `fakeTimer.uninstall()` added (`restore` was/is an alias of this all along)
- [x] `FakeXMLHttpRequest` cleaned up, some statics were incorrectly on the prototype instead
- [x] all-optional interfaces replaced with `Partial<T>`
- [x] `hasNested`, `every` and `some` matchers added
- [x] sandbox config types tightened
- [x] `sinon.fake` added (see [here](http://sinonjs.org/releases/v5.0.3/fakes/))
- [x] `sandbox.replace` added
- [x] `sandbox.replaceGetter` added
- [x] `sandbox.replaceSetter` added
- [x] **sinon static/namespace now inherits sandbox**
- [x] legacy sandbox interface added
- [x] sinon namespace types cleaned up and tightened
- [x] conditional types (TypeScript 2.8) used for stubbed instances (#25656)

## Misc

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: http://sinonjs.org/releases/v5.0.3/
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by 43081j** [comment]
questions:

* do we want to roll the TS version requirement back and remove the conditional types? otherwise all types depending on sinon (a lot do for tests) need to be updated to also require 2.8
* we have an issue where `const Sinon` and `namespace Sinon` now seem to conflict, im not sure why they didn't before

**Comment by andy-ms** [comment]
With the exception of `hexo-log` (#25511) and `mochaccino`, the dependers on `sinon` all have `sinon` in their name, so you'd be updating the whole sinon ecosystem at once.
However, this is a pretty popular package ... would it be possible to replace `SinonStubbedMember<T>` with `SinonStub & T`? Not accurate but should allow code to compile?

**Comment by 43081j** [comment]
@andy-ms the ones in my original post all seem to `require` sinon but most of them are using it for their tests (not in their own types). still, it would mean all that depend on those will need bumping too.

the best i can do in regards to the conditional type i think is to roll back to:

```
[P in keyof TType]: SinonStub
```

This just means we reintroduce the slightly-wrong types we had before, where primitive properties end up as `SinonStub` even though they are unchanged.

So:

```
const foo = {
  a: 1,
  b: 2,
  c() { return 3; }
};
const bar = createStubInstance(foo);
bar.a; // SinonStub but is really a number still
bar.b; // SinonStub but is really a number still
bar.c; // SinonStub, as it should be
```

`SinonStub & T` (where `T` is the type of the member) would result in pretty much the same situation really, maybe worse/more confusing.

**Comment by andy-ms** [comment]
If something only uses `sinon` in its tests, those tests should probably just be changed to not depend on it like I did in #25511. But if it has `sinon` in the name I would presume the dependency is legitimate...

Still, even if the only affected package were `sinon`, it's probably better to hold off on requiring 2.8 yet since this is a commonly-required package and ts2.8 is only a month old -- maybe change it in a few months.

**Comment by 43081j** [comment]
thats fair enough, i can change it back to what it was before (people seemed fine with that before).

as for the packages.. i found those with a quick grep for anything referencing `sinon` then filtered down to only those which require/import it.  ill take a look at each one later and see if they need updating/cleaning up.

**Comment by 43081j** [comment]
@andy-ms ill be adding a bunch of tests first too, will comment again once they're pushed.

we're missing quite a few and i may as well clean the existing ones up a bit.

**Comment by 43081j** [comment]
alright so, i cleaned up some incorrect types and wrongly placed ones.

then i rewrote the whole test file. it would be nice if someone can tell me my tests are wrong/terrible/horrendous or not. im happy to roll back to the old ones but, unless i misunderstood everything, these should be better?

do let me know, whoever has chance to review

**Comment by 43081j** [comment]
Please don't merge yet.

This needs a second pair of eyes, whether they are a contributor of this particular package or not. It wouldn't be right to merge based on my own (author) review.

Ideally, I need someone to review the tests and changes as a whole (maybe try it out).

**Comment by 43081j** [comment]
i squashed and rebased onto master FYI.

the 2.8-specific types got merged after all in the other, smaller PR.

**Comment by andy-ms** [comment]
@JoshuaKGoldberg Could you review?

**Comment by 43081j** [comment]
alright here we go

@andy-ms @JoshuaKGoldberg  i have rebased onto master (where TS2.8 changes are reverted) and have cherry-picked #25656 into here.

all should be mergeable _and_ contain the conditional types now.

File: types/sinon/sinon-tests.ts
Comment (Flarna) [inline types/sinon/sinon-tests.ts]: There are no tests for the inherited `Spy` methods. Maybe add something like `const spy: sinon.SinonSpy = stub` to verify that they match?

**Comment by Flarna** [comment]
I used your the typings from this PR to port my project to Sinon v5 and it worked fine. It's a Node project and I don't use all Sinon features.
So for me this looks fine.

**Comment by 43081j** [comment]
@Flarna much appreciated.

i understand this is a crazily big PR in that pretty much most of the types have changed, moved, etc. So it is rather difficult to review on a "diff-by-diff" basis.

The only way we can really be sure all is fine i suppose, is by having a look through the final definitions, trying it out and ensuring the tests pass (they do).

So thanks for trying them out on your project.

File: types/sinon/index.d.ts
Comment (43081j) [reply types/sinon/index.d.ts]: good catch! you are correct, i had missed that when de-duping these

File: types/sinon/index.d.ts
Comment (43081j) [reply types/sinon/index.d.ts]: yup that makes sense, will get that added

File: types/sinon/sinon-tests.ts
Comment (43081j) [reply types/sinon/sinon-tests.ts]: 👍  we don't need to test each individual member, just that the interface is inherited, so you are correct

**Comment by 43081j** [comment]
i've pushed the changes from the code review now.

everything seems fine.

File: types/sinon/index.d.ts
Comment (43081j) [reply types/sinon/index.d.ts]: yup my bad, my editorconfig got mixed up, and the jsdoc should be on the function rather than the interface in fact

**Comment by 43081j** [comment]
@Flarna  mind giving the latest push a go? moved the deprecation to the method and fixed the whitespace mixup

**Review by Flarna — APPROVED** [review]

**Comment by Flarna** [comment]
@43081j Still works for me and correctly issues deprecation warnings. Besides that I like the changes in tests as now they straight forward verify the type definitions and not sinon => a clear ship it from my point of view.

**Comment by 43081j** [comment]
👍 thanks very much for reviewing and trying it out in your project, its been a fairly big PR that wasn't so easy to review.

and yes hopefully this means the tests are much cleaner now. in future its far clearer where to test part of the API rather than having everything mixed up.

**Comment by 43081j** [comment]
@andy-ms think this is ready to be reviewed now/merged?

**Comment by Harrison-M** [comment]
Haven't been able to check out all the nooks and crannies of this just yet, but I'm also using this on a current project and it's compiling fine. In partciular, the new `sinon.fake()` stuff is working for me.

**Comment by 43081j** [comment]
@andy-ms any chance we can get this merged? if its just waiting because theres plenty of PRs to review, no worries, just want to make sure it hasn't been lost/tagged weirdly.

File: types/sinon/index.d.ts
Comment (UselessPickles) [inline types/sinon/index.d.ts]: Should the restore() method also be moved to SinonFakeXMLHttpRequestStatic?

Comment (43081j) [reply types/sinon/index.d.ts]: @UselessPickles you would be correct it seems. this was leftover from what we had before, which was also wrong back then! 

looks to me like sinon's fake XHR has `restore` as a static method, _not_ as an instance method, so this doesn't belong here at all. it belongs only on the static interface

Comment (UselessPickles) [reply types/sinon/index.d.ts]: Good to know I'm not insane :). This caused compiler errors for me, but I wasn't confident enough to submit my own PR for it.

Comment (43081j) [reply types/sinon/index.d.ts]: @UselessPickles  sorry if you had started on that, i just threw one up at #26119

**Pull request merged**
